linkerd/app/core/src/errors.rs

```diff
@@ -159,6 +239,11 @@ fn set_grpc_status(error: &Error, headers: &mut http::HeaderMap) -> grpc::Code {
             headers.insert(GRPC_MESSAGE, msg);
         }
         code
+    } else if error.is::<hyper::error::Error>() {
```

Comment:
That seems super naive... However, most of the methods on hyper::Error are private and I could not figure out a way to determine that this error is `Error(Body, Error { kind: Io(Kind(BrokenPipe)) })`

Reply:
I think we can actually use `hyper::Error::is_closed` for this: https://github.com/hyperium/hyper/blob/9a8413d91081ad5a949276f05337e984c455e251/src/error.rs#L114-L129

Reply:
Though, we'd have to check a few other methods besides that one. Perhaps we can use [`Error::into_cause`](https://github.com/hyperium/hyper/blob/9a8413d91081ad5a949276f05337e984c455e251/src/error.rs#L145-L148)?

Reply:
Not sure we can. In this particular case we get a kind of `Io`, so right here https://github.com/hyperium/hyper/blob/9a8413d91081ad5a949276f05337e984c455e251/src/error.rs#L317. If we to into_cause, this does not give us much apart from being able to  look at the textual form of this error. Not sure we want to be comparing strings.

Reply:
(As we discussed offline...) It looks like `into_cause` should be able to return the underlying `io::Error`.

Reply:
@olix0r So I am a bit confused here. into_cause consumes the error and gives you the optional cause. Do we actually want to use that? Dont we want something that just takes it by ref ()judging by the signature of the `set_grpc_status` func. Alternatively we can use Error::source but that gives me back a `Option<&(dyn StdError + 'static)>` which we cannot pipe recursively into `set_grpc_status` as it is not `Send + Sync` (correct me if I am wrong). Looks like we want to use cause2 which gives back `Option<&(dyn StdError + 'static + Sync + Send)>` but that... is deprecated. Not sure what is the best option here.  Also, what is the reason for `set_grpc_status` to take an `&Error` instead of `Error`. Is it because of the `ServiceError` holding an `Arc` to its inner error ?

Reply:
Y'know... this is all probably overcomplicating it. I think it's fine to just match `hyper::Error` for now. We should leave a comment explaining the situation...

Reply:
But still what would be a proper way to introspect that error type according to you ?

Reply:
After looking more closely, I believe we can probalby modify `set_grpc_status` to consume the error... We'll have to change some of those `downcast_ref` to `downcast`.

Reply:
Will give it a shot

Reply:
The main head scratcher for me is the fact that we are handling a `ServiceError` here, that holds an Arc to its inner error. So we can only take a shared ref of it and cannot consume it, thereby preventing us from calling the method recursively. Can we at that point of the code assume that .`try_unwrap `on the Arc holding the inner error will succeed ?

Reply:
No, that won't work, because the error is shared with the lock/buffer--both of these layers are used to share an inner service, and when its inner service fails, we have to share that error with all outer instances... So try_unwrap won't have exclusive ownership of the error.

As I said, I'm fine with just matching on hyper::Error in this case to get this to master. I'd rather get the fix out and refine it on our own time. I'll take a quick look at the error handler and see if there's a way to simplify the problem, but I'm happy to submit that as a followup. If I can figure it out, that is.

Reply:
Honestly, the best option might be to submit a PR to hyper to add an `Error::cause` like

```rust
    /// Returns a reference to the cause
    pub fn cause(&self) -> Option<&Box<dyn StdError + Send + Sync>> {
        &self.inner.cause
    }
```

We probably won't pick up this change until tokio 0.2 merges, but it would allow us to clean this up later...